src/backend/access/appendonly/appendonlywriter.c

```diff
@@ -778,6 +778,53 @@ RegisterSegnoForCompactionDrop(Oid relid, List *compactedSegmentFileList)
 	return;
 }
 
+void
+UpdateSegnoAfterCompaction(Oid relid, List *compactedSegmentFileList)
+{
+	AORelHashEntryData *aoentry;
+	Relation	aosegrel;
+	bool	   *awaiting_drop = NULL;
+	int			i;
+
+	Assert(Gp_role != GP_ROLE_EXECUTE);
+	if (Gp_role == GP_ROLE_UTILITY)
+	{
+		return;
+	}
+
+	if (compactedSegmentFileList == NIL)
+	{
+		return;
+	}
+
+	aosegrel = relation_open(relid, AccessShareLock);
+	awaiting_drop = get_awaiting_drop_status_from_segments(aosegrel);
+	relation_close(aosegrel, AccessShareLock);
+
+	acquire_lightweight_lock();
+
+	aoentry = AORelGetOrCreateHashEntry(relid);
+	Assert(aoentry);
+
+	for (i = 0; i < MAX_AOREL_CONCURRENCY; i++)
```

Comment:
Is it possible to use foreach over compactedSegmentFileList instead of this loop? It seems that it will be faster

Reply:
I just copied RegisterSegnoForCompactionDrop and modified a bit.

Reply:
fixed